tests/functional/test_stub.py

```diff
@@ -0,0 +1,122 @@
+
+import botocore
+from botocore.stub import Stubber
+from botocore.credentials import Credentials
+from botocore.exceptions import StubResponseError, ClientError
+from botocore.loaders import Loader
+from botocore import hooks
+import botocore.client
+import botocore.retryhandler
+import botocore.translate
+
+
+class TestStubber(unittest.TestCase):
+    def setUp(self):
+        credentials = Credentials('access-key', 'secret-key')
+
+        with mock.patch('botocore.client.EndpointCreator') \
+                as endpoint_creator_patch:
+            endpoint_creator = endpoint_creator_patch.return_value
+            self.endpoint = mock.Mock()
+            self.endpoint.host = 'https://s3.amazonaws.com'
+            self.endpoint.make_request.return_value = (
+                mock.Mock(status_code=200), {})
+            endpoint_creator.create_endpoint.return_value = self.endpoint
+
+            resolver = mock.Mock()
+            resolver.construct_endpoint.return_value = {
+                'properties': {},
+                'uri': 'http://foo'
+            }
+
+            event_emitter = hooks.HierarchicalEmitter()
+            client_creator = botocore.client.ClientCreator(
+                Loader(), resolver, 'user-agent', event_emitter,
+                botocore.retryhandler, botocore.translate, None)
+
+            config = botocore.client.Config(signature_version=botocore.UNSIGNED)
+
+            self.client = client_creator.create_client(
+                's3', 'us-west-2', credentials=credentials,
+                client_config=config)
+
+        self.stubber = Stubber(self.client)
+
+        self.parser_patch = mock.patch(
+            'botocore.endpoint.parsers.ResponseParser.parse')
+        self.parser_patch.start()
+
+    def tearDown(self):
+        self.parser_patch.stop()
+
+    def test_stubber_prevents_http_call(self):
+        self.stubber.add_response('list_objects', {})
+        self.stubber.activate()
+        self.client.list_objects(Bucket='foo')
+        self.assertFalse(self.endpoint.make_request.called)
+
+    def test_stubber_returns_response(self):
+        service_response = {'ResponseMetadata': {'foo': 'bar'}}
+        self.stubber.add_response('list_objects', service_response)
+        self.stubber.activate()
+        response = self.client.list_objects(Bucket='foo')
+        self.assertEqual(response, service_response)
+
+    def test_non_activated_stubber_does_nothing(self):
+        # Make sure the request is sent properly when nothing has been done.
+        self.client.list_objects(Bucket='foo')
+        self.assertTrue(self.endpoint.make_request.called)
+        self.endpoint.make_request.reset_mock()
+
+        self.stubber.add_response('list_objects', {})
+        self.client.list_objects(Bucket='foo')
+        self.assertTrue(self.endpoint.make_request.called)
+
+    def test_deactivated_stubber_does_nothing(self):
+        self.stubber.add_response('list_objects', {})
+        self.stubber.activate()
```

Comment:
Now that I think about it. I would even suggest moving a lot of these tests to the unit directory if the purpose of the test is to see if the endpoint was hit. Then, in the functional test, it could be a some smoke cases that ensure the proper responses and errors are returned and the external interface is as you would expect.